### DISCUSSION TITLE
LocalExecutor run Async operators without a triggerer process

### DISCUSSION THREAD

**Discussion by raphaelauv** [apache/airflow#24606] [General]
### Description

Currently the LocalExecutor can't run Async operators, it need a triggerer process 

### Use case/motivation

LocalExecutor "should" be a fully airflow working process without needing another process to run tasks

It would keep airflow simple for entry users and small deployments ( and even big deployments if using only async operators).

----

The simplest deployment with async support is currently this -> 

![localexectur (1)](https://user-images.githubusercontent.com/10202690/175033423-cf060f66-c459-4b1a-8c75-6950c28a65e3.jpg)

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by uranusjr** [comment]
This sounds nice, but quite impossible unless we either put the triggerer loop in a thread/subprocess (not easy), or rewrite the entire executor in async (much harder). It’d be a cool project to work on though.

**Comment by mik-laj** [comment]
What do you think about `airflow standalone`command? Is it meer your requirements?

**Comment by raphaelauv** [comment]
`airflow standalone` is more for starting a local airflow , nothing close to a production setup.

`LocalExecutor` is perfectly prod ready and enough for deployment that are only running operators triggering operations in external solutions and not having needs for elastic scaling ( because not big amount of airflow tasks )

**Comment by potiuk** [comment]
I think it's wrong. We should instead make standalone a fully-production ready solution. Merging Triggerer into Scheduler is a bad idea.

Converting it into discussion.

**Reply by potiuk** [reply]
BTW. Airflow Standalone in this "production mode" SHOULD run Local Executor and SHOULD allow to run tasks in parallel. There are two options we could do it BTW:

a) make airlfow integration better with sqlite to allow parallelism (the fact that currrently only Sequential Executor can be used with SQLite is not a limitation of SQLite but our way of integration with it)

b) allow `airflow standalone` to connect to external DB (and use Local Executor)

**Reply by potiuk** [reply]
BTW. The a) oprion is very appealing. 

It would have the added benefit that you'd not need anything to run airlfow, in a small production mode

* `pip install airflow`
* `airflow standalone` 

And you are done.

**Reply by raphaelauv** [reply]
I'm not proposing to ```Merging Triggerer into Scheduler`

I'm proposing to add an async loop to the LocalExecutor so it can run async operators.

---

about `option a)` a production mode without an external DB would be strange from my point of view

**Reply by potiuk** [reply]
> I'm proposing to add an async loop to the LocalExecutor so it can run async operators.

This is merging Triggerer into Scheduler actually. There is much more to triggerer thatn "async loop" - there are a number of database operations, there is startup and recovery process etc. etc.

> a production mode without an external DB would be strange from my point of view

I think it depends on how you treat SQLIte. SQLIte is actually a fully fledged database, with "serverless" local file storage. If you use LocalExecutor in Production (which by definition runs on a single machine and has no redundancy or anything like that) I wonder why do you think not having an "external" DB is strange in this case:

* you can back it up easily (copy file) and restore (copy file back), You can store in RAID-ARRAY backed filesystem, you can also keep it on remote networking storage for additional redundancy if you ave local network. With some filesystems (ZFS) you even have snapshots that will allow you to travel-in-time with the changes in the DB (which most of the commercial databases do not have)
* you can connect and query it with SQL client
* For airflow the DB is really "interna" detail and the way to store meta-data. That's basically it. We do not need multi-processor, efficient data manipulation, transformations, heavy databases to run heavy processing when you are using Local Executor and single machine. 
* SQLIte is actually used for small "local" databases in production by multiple buisnesses - it is the best server-less production-level DB there is.

What makes you think it's strange? What would you miss if you had Internal Metadatada Sqlite database as your medatada db @raphaelauv ? Could you be more specific what the problems are ? Or is it just "perception"?

**Reply by raphaelauv** [reply]
thanks @potiuk for your answers, I'm always learning a little bit about airflow thanks to you ( I just hope my questions are not too boring for you :sweat_smile:  )

I'm not arguing against `SQLIte` itself , more about not having the possibility to have an external database like an already deployed database server (multiple DB on same server) or a new one ( and in both case cloud managed or not ) with `airflow standalone`

also `LocalExecutor` is compatible with `HA` scheduler but you need a `PostgreSQL 10+` or `MySQL 8+` for that

----

`This is merging Triggerer into Scheduler actually. `

Okay then it's a way more complex topic than I thought. 

My motivation here is/was to find the simplest design ( so without using the CeleryExecutor or KubernetesExecutor ) that is HA and compatible with async operators

**Reply by potiuk** [reply]
> I'm not arguing against SQLIte itself , more about not having the possibility to have an external database like an already deployed database server (multiple DB on same server) or a new one ( and in both case cloud managed or not ) with airflow standalone

I think we have a perfect agreement here. `airflow standalone` that we are talking here COULD have a way to connect to Postgres/MySQL. No problem whatsoever with this. HA scheduler is not  a reason though. When you want HA scheduler then you are already falling into the camp of running complex deployment where you have multiple components. And if you do, then the benefits of having separately deployable components (for example running 2 schedulers and 5 triggerers if you want) far outweigh the convenience of having multiple, independent components put together in a single binary. 

Genarally speaking I think there are only two possible ways of "production" deployment (and they are targetting different kinds of users);

1) 'airflow standalone' - for a small team/single person managing scheduling in their production environment. There is a huge benefit of having everything "bundled" in a single "command/binary/package" that you can install Airflow in a single  `pip install` opreation and have single `serviced` or similar to start and run it - no cross-component management, no docker image needed (though it could be run as a container as well) no docker-compose and managing cross-component configuration. This **might** use sqlite or it **might** connect to a database. In many small teams, the company/db-admins might still provide a "database-as-a-service" that you should be able to treat as a "given".

2) Configurable deployment with multiple components managed separately - scheduler, webserver, trigger, dag-file-processor - in the future internal-api server. Once you have the need to run more then one component, you immediately fall into this camp - when you have 2 scheduler,s webserver, database, already as separate components you have to have some way to "coordinate them" (helm chart, docker-compose, swarm, custom kubernetes deployment) - and in this case adding triggerer as separate components makes much more sense than trying to embed it in another component. This is really following the modern "production" deployment paradigms that each running component has separate container and runs independently from the others. It's not yet "microservice" approach (which I personally think went too much "micro") but it's using the current best practices of reaping the benefits of being able to run, manage, restart, configure each isolated component separately. It comes with cost/complexity, but once you have two -three components that talk to each other and are isolated, you've already paid the cost and adding one more components to your deployment costs almost nothing. 

I think merging **just** triggerer to scheduler without merging all other components (which `airflow scheduler` does effectively`) is basically merging those two approach and get worst of both worlds - you still have to maintain complexity of cross-component management, but you have some "logical" component merged logically with another with no apparent reason or benefit - and it make it more complex to manage them as they both have to be manged and scaled together. There is really no reason why there should be the same number of schedulers as triggerers for example. Those two have completley different characteristics and "load" of both components are pretty different. Scheduler mostly runs heavy database queries, where triggerer is running mostly async-io loop which - when we have more operators converted to it - effectively what it will do is just "occasionally" read/write from the database but most of the work will be to open insane amount of sockets to external systems and have an async-io loop waiting for incoming messages on those sockets and running small snippets of python code when something arrives to determine "are we there yet" for thousands of deferred tasks. This basically means that Triggered will get completely different expectations about the resource usage than scheduler and by separating it out you will be able to manage them separately and optimise it much better.

I can very easily imagine that we have 2 schedulers and 4 triggerers running (if you run super async-io-heavy load) and also the opposite - 4 schedulers and 1 triggerer to handle an occasional asyncio operation. By merging the two we are not achieving anything else than effectively limiting the flexibility and always assuming 1 scheduler = 1 trigger. It very slightly decreases the operational effort but heavily limits the flexibility.

**Reply by potiuk** [reply]
Also I think the 1) are 2) are really what we should target with the discussion started by @vikramkoka https://lists.apache.org/thread/smfpovyh1hfych9wmbsnvwt397hh7g11 : For me the "standalone" is really best for the 1) user segment that Vikram mentioned, where "componentised airflow" is perfect for the 2) segment of users.

**Reply by raphaelauv** [reply]
Thanks for you detailed answer , it make perfect sense to keep theses components separate so deployments can scale them separately.

Also add to the standalone command the possibility to connect to an external database would be a useful feature.

**Comment by raphaelauv** [comment]
@uranusjr , I created a LocalAsyncExecutor starting a thread running a triggerer 


```python
class ExecutorLoader:
    """Keeps constants for all the currently available executors."""

    _default_executor: Optional["BaseExecutor"] = None
    executors = {
        LOCAL_EXECUTOR: 'airflow.executors.local_executor.LocalExecutor',
        "LocalAsyncExecutor": 'airflow.executors.local_async_executor.LocalAsyncExecutor',
        
        ...
```

```python
from threading import Thread
from typing import Optional

from airflow.executors.local_executor import LocalExecutor
from airflow.exceptions import AirflowException
from airflow.jobs.triggerer_job import TriggererJob


class LocalAsyncExecutor(LocalExecutor):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        self.thread_triggerer_job: Optional[Thread] = None
        self.triggerer_job: Optional[TriggererJob] = None

    def start(self) -> None:
        """Starts the executor"""
        self.triggerer_job = TriggererJob()
        self.thread_triggerer_job = Thread(target=self.triggerer_job.run)
        self.thread_triggerer_job.start()
        super().start()

    def sync(self) -> None:
        if not self.thread_triggerer_job.is_alive():
            raise AirflowException("triggerer process no more running")
        super().sync()

    def end(self) -> None:
        """Ends the executor"""
        self.triggerer_job.on_kill()
        self.thread_triggerer_job.join()
        super().end()

```

it run and also gracefully close

```
{scheduler_job.py:179} INFO - Exiting gracefully upon receiving signal 15
{process_utils.py:125} INFO - Sending Signals.SIGTERM to group 149. PIDs of all processes in the group: [149]
{process_utils.py:80} INFO - Sending the signal Signals.SIGTERM to group 149
{process_utils.py:75} INFO - Process psutil.Process(pid=149, status='terminated', exitcode=0, started='18:43:51') (149) terminated with exit code 0
{local_executor.py:398} INFO - Shutting down LocalExecutor; waiting for running tasks to finish.  Signal again if you don't want to wait.
{triggerer_job.py:111} INFO - Waiting for triggers to clean up
{triggerer_job.py:117} INFO - Exited trigger loop
{process_utils.py:125} INFO - Sending Signals.SIGTERM to group 149. PIDs of all processes in the group: []
{process_utils.py:80} INFO - Sending the signal Signals.SIGTERM to group 149
{process_utils.py:94} INFO - Sending the signal Signals.SIGTERM to process 149 as process group is missing.
{scheduler_job.py:780} INFO - Exited execute loop
[9] [INFO] Handling signal: term
[30] [INFO] Worker exiting (pid: 30)
[10] [INFO] Worker exiting (pid: 10)
[9] [INFO] Shutting down: Master

```
